### ISSUE TITLE
Wrong red LED colour for WS2812

### ISSUE THREAD

**Issue by flashyas1** [FastLED/FastLED#90]
I'm working with a WS2812 strip and have tested it using the RGBCallibrate example program and the red LEDs are orange in FastLED V3.0.0. 

They appear fine when tested with the Neopixel library. I then tested an old version of FastLED from http://www.tweaking4all.com/hardware/arduino/arduino-ws2812-led/ and the issue was fixed when running the same RGBCallibrate program.

**Comment by grassm** [comment]
I have the same problem.   It is a hardware problem from the stand point that some of the WS2812 chips show green when you tell it to show red.   You can fix it by changing the color order from RGB to GRB.  I've not come across any WS2812 that the blue isn't correct.   So far.....if I buy the 30/m or 60/m I have to use the GRB sequence to get red and green to sequence properly.  When I buy the 144/m  I have to use the RGB sequence to get red as red and green as green.

**Comment by focalintent** [comment]
@grassm it's not a hardware problem - it's just that different manufacturers use different data ordering - some use RGB, some use GRB, some use BGR - this is why the library allows you to change the color ordering when adding leds.

**Comment by ehsanhaleem** [comment]
Grassm
You save my life thank you

**Comment by grassm** [comment]
@focalintent   Hardware vs software problem.   The difference IS THE HARDWARE!  It's not something wrong with the software(ide, library, etc.), it is a difference between WS2812 hardware.  Let's say we're buying AA cells and I buy some and the button end is + and the next ones I buy the button is -. Clearly the ones with the negative button end are wrong.  It's a hardware problem.   The same goes for any hardware part.   2n2222, IRF3408, etc.  The PROBLEM is the hardware isn't the same, so it is a HW problem.

**Comment by grassm** [comment]
@ehsanhaleem Well, I don't think I saved your life, but if I did, then you are welcome.

**Comment by focalintent** [comment]
It's a hardware consistency issue - but there's no standard way to order the color data for the led chips/chipsets - in fact, this is _**exactly why**_ i put support in for changing the color ordering.  But it's not something "wrong" with the hardware - it's more akin to the difference between cars that put the headlight controls on the left hand side of the steering wheel vs. cars that put them on the right.  So - there's nothing wrong with the hardware - it's just done differently.  Problem implies that something is wrong/broken.

**Comment by grassm** [comment]
Well, I'd agree with you except for the fact that WS2812 is to addressable LEDs as  a   AA cell is to batteries/cells.  The AA is a specific form, fit and function.  If you buy a car battery say group 27F,  it has a certain Form, Fit, and Function. It should work in any vehicle that calls for that battery, without any modification.   If I buy an IRFXXXX MOSFET from any manufacture, I'd expect it to work in any circuit that calls for it (without modification).  
Please don't misunderstand.  I am so thankful for everyone like you that takes the time to create these libraries. I am a very big fan of standardization.  Yes, sometimes things have to change, but if they are not going to be specific in the order of RGB, then they need to add a field to WS2812B.........WS2812B-RGB,  or  WS2812B-GRB   WS2812B-BRG,   or  WS2812B-BGR.  When you order a car,  you need to tell them what color you want, etc, but the headlights will always be on the front, taillights on the back.  Maybe a better analogy would be with TVs.   When you buy a TV,  the RGB of the tv is always just that, RGB!


   
--------------------------------------------
On Mon, 10/23/17, Daniel Garcia <notifications@github.com> wrote:

 Subject: Re: [FastLED/FastLED] Wrong red LED colour for WS2812 (#90)
 To: "FastLED/FastLED" <FastLED@noreply.github.com>
 Cc: "Mark Grass" <fairwindsii@yahoo.com>, "Mention" <mention@noreply.github.com>
 Date: Monday, October 23, 2017, 9:05 PM
 
 It's a hardware consistency
 issue - but there's no standard way to order the color
 data for the led chips/chipsets - in fact, this is
 exactly why i put support in for
 changing the color ordering.  But it's not something
 "wrong" with the hardware - it's more akin to
 the difference between cars that put the headlight controls
 on the left hand side of the steering wheel vs. cars that
 put them on the right.  So - there's nothing wrong with
 the hardware - it's just done differently.  Problem
 implies that something is wrong/broken.
 
 —
 You are receiving this because you were mentioned.
 Reply to this email directly, view
 it on GitHub, or mute
 the thread.

**Comment by focalintent** [comment]
Good luck getting any of the led manufacturers/factories to agree with you :)  This was the case years before the WS2812 came out, and will likely continue to be the case for years to come.

Some of it comes from the controller chips themselves - all they know is they take in data, and use that data to drive PWM on 3 to 9 outputs (the TM1809 would drive 3 RGB leds from a single controller chip, the LPD8806 would drive two from one).  What pins went to what color led depended heavily on the form factor the chips and leds were being put onto and how things were wired up, and so one led strip might have pin 0 wired to red, pin 1 to green, and pin 2 to blue, while another strip might have a different ordering, etc... etc... For that matter, i've had FastLED users who buy their own WS2811 controller chips and wire up the rgb leds to the chips on their own - and they can use whatever ordering they want! (hopefully it's consistent across their project, though :)

Also, WS2812B is almost a meaningless identifier these days - I've seen a number of led strips that are sold as WS2812 that, while they use the same protocol, are actually different (e.g. SK6812) - both in terms of ordering of RGB, but also in the exact timings being used.   (I also once saw an APA102 clone being sold as WS2812B compatible).  And then there's the fact that World Semi themselves _changed the WS2812B_ but are still calling it WS2812B.  Yippie!

To play off of your AA battery analogy - the situation with buying LEDs is more like you are buying something that is AA sized for a toy that takes 4 of them, and you just grab something based on the size and fit because it's cheap and delivers quickly.  However, if you don't look closely enough at it, you may not realizing you're ending up with something wildly different, e.g. a Li-Ion 3.7v battery instead of the AA 1.2-1.5 volt battery you were expecting[1].  Toss four of those into your AA powered toy, and you're pumping nearly 15v through it instead of the 5-6v that it may be expecting.

So, basically, no matter what you get - do a quick test before you permanently wire/connect things up, so you know what you're putting in place.  (You should be doing that anyway, to make sure you don't have any chips that are DOA - trust but verify and all that jazz)

[1] yes, these are real - https://www.amazon.com/Tenergy-Li-Ion-Cylindrical-Rechargeable-800mAh/dp/B003BHDP4C

**Comment by just4phil** [comment]
i have the same problem ... using an teensy++2 ..... if i try to make a green color then it looks like orange .... red and blue work like a charm ..strange

**Comment by KTibow** [comment]
Just to clear this up:  
Wherever you put `RGB` or `GBR` or `BGR` while you were initializing the LEDs:  
Try the other two combos.

**Comment by ddaydd** [comment]
@KTibow  Thanks, same problem, this worked for me

`  FastLED.addLeds<WS2812B, PIN_LEDS, GRB>(leds, NUM_LEDS);`

**Comment by ChukasNeves** [comment]
although the WS2812B expects the Color Order GRB:
`FastLED.addLeds<WS2812B, PIN_LEDS, GRB>(leds, NUM_LEDS);`



if I do: `fill_solid(leds, NUMMATRIX,CRGB(255, 0 , 0) ); //expect Red color `
The result is a Green color. 
This is fixed by setting the desired color sequence to RGB:
`FastLED.addLeds<WS2812B, PIN_LEDS, RGB>(leds, NUM_LEDS);`

now `fill_solid(leds, NUMMATRIX,CRGB(255, 0 , 0) ); //expect Red color`  outputs the color red
also colors `CRGB( 0, 255 , 0)` (green) and `CRGB( 0 , 0, 255)` (blue) function properly this way.
